File: nodejs/backend/src/repository/LeadRequestRepository.ts
import { Pool } from 'pg';
import { LeadRequest } from '../model/LeadRequest';

function rowToLeadRequest(row: any): LeadRequest {
  return {
    id: row.id,
    name: row.name,
    email: row.email,
    phone: row.phone,
    service: row.service,
    message: row.message,
    createdAt: row.created_at,
  };
}

export class LeadRequestRepository {
  constructor(private readonly pool: Pool) {}

  async insert(request: LeadRequest): Promise<void> {
    await this.pool.query(
      `INSERT INTO lead_requests (id, name, email, phone, service, message, created_at)
       VALUES ($1, $2, $3, $4, $5, $6, $7)`,
      [
        request.id,
        request.name,
        request.email,
        request.phone,
        request.service,
        request.message,
        request.createdAt,
      ],
    );
  }

  async findAll(): Promise<LeadRequest[]> {
    const result = await this.pool.query(
      `SELECT id, name, email, phone, service, message, created_at
       FROM lead_requests ORDER BY created_at DESC`,
    );
    return result.rows.map(rowToLeadRequest);
  }

  async deleteById(id: string): Promise<LeadRequest | null> {
    const result = await this.pool.query(
      `DELETE FROM lead_requests WHERE id = $1
       RETURNING id, name, email, phone, service, message, created_at`,
      [id],
    );
    return result.rows[0] ? rowToLeadRequest(result.rows[0]) : null;
  }
}

File: rust/backend/src/repository/lead_request_repository.rs
use crate::model::LeadRequest;
use async_trait::async_trait;
use sqlx::PgPool;
use uuid::Uuid;

#[async_trait]
pub trait LeadRequestRepository: Send + Sync {
    async fn insert(&self, request: &LeadRequest) -> Result<(), sqlx::Error>;
    async fn find_all(&self) -> Result<Vec<LeadRequest>, sqlx::Error>;
    async fn delete(&self, id: Uuid) -> Result<Option<LeadRequest>, sqlx::Error>;
}

#[derive(Clone)]
pub struct PgLeadRequestRepository {
    pool: PgPool,
}

impl PgLeadRequestRepository {
    pub fn new(pool: PgPool) -> Self {
        Self { pool }
    }
}

#[async_trait]
impl LeadRequestRepository for PgLeadRequestRepository {
    async fn insert(&self, request: &LeadRequest) -> Result<(), sqlx::Error> {
        sqlx::query(
            "INSERT INTO lead_requests (id, name, email, phone, service, message, created_at)
             VALUES ($1, $2, $3, $4, $5, $6, $7)",
        )
        .bind(request.id)
        .bind(&request.name)
        .bind(&request.email)
        .bind(&request.phone)
        .bind(&request.service)
        .bind(&request.message)
        .bind(request.created_at)
        .execute(&self.pool)
        .await?;

        Ok(())
    }

    async fn find_all(&self) -> Result<Vec<LeadRequest>, sqlx::Error> {
        sqlx::query_as::<_, LeadRequest>(
            "SELECT id, name, email, phone, service, message, created_at
             FROM lead_requests ORDER BY created_at DESC",
        )
        .fetch_all(&self.pool)
        .await
    }

    async fn delete(&self, id: Uuid) -> Result<Option<LeadRequest>, sqlx::Error> {
        sqlx::query_as::<_, LeadRequest>(
            "DELETE FROM lead_requests WHERE id = $1
             RETURNING id, name, email, phone, service, message, created_at",
        )
        .bind(id)
        .fetch_optional(&self.pool)
        .await
    }
}
